internal/engine/buildcontroller.go

```diff
@@ -32,23 +33,73 @@ func NewBuildController(b BuildAndDeployer) *BuildController {
 	}
 }
 
+// Algorithm to choose a manifest to build next.
+func nextManifestToBuild(state store.EngineState) model.ManifestName {
+	// First, go through all the manifests in order.
+	// If any of them haven't started yet, build them now.
+	for _, mn := range state.ManifestDefinitionOrder {
+		ms, ok := state.ManifestStates[mn]
+		if ok && !ms.StartedFirstBuild {
+			return mn
+		}
+	}
+
+	// Next go through all the manifests, and check:
+	// 1) all pending file changes, and
+	// 2) all pending manifest changes
+	// The earliest one is the one we want.
+	choiceName := model.ManifestName("")
+	earliest := time.Now()
+
+	// always use a stable iteration order
+	for _, mn := range state.ManifestDefinitionOrder {
+		ms, ok := state.ManifestStates[mn]
+		if !ok {
+			continue
+		}
+
+		// Always prioritize builds that are crash-rebuilding.
+		if ms.CrashRebuildInProg {
+			return mn
+		}
+
+		t := ms.PendingManifestChange
+		if t.Before(earliest) && ms.IsPendingTime(t) {
+			choiceName = mn
+			earliest = t
+		}
+
+		spurious, _ := onlySpuriousChanges(ms.PendingFileChanges)
+		if !spurious {
+			for _, t := range ms.PendingFileChanges {
+				if t.Before(earliest) && ms.IsPendingTime(t) {
+					choiceName = mn
+					earliest = t
+				}
+			}
+		}
+	}
+
+	return choiceName
+}
+
 func (c *BuildController) needsBuild(ctx context.Context, st store.RStore) (buildEntry, bool) {
 	state := st.RLockState()
 	defer st.RUnlockState()
 
-	if len(state.ManifestsToBuild) == 0 {
+	if c.lastActionCount == state.BuildControllerActionCount {
```

Comment:
Can you explain this line? i found it confusing when I was last in this code

Reply:
added

```
	// Don't start the next build until the previous action has been recorded,
	// so that we don't accidentally repeat the same build.
```